### ISSUE TITLE
Integration and Test Infrastructure at NERSC

### ISSUE THREAD

**Issue by weaverba137** [desihub/desiutil#29]
This was originally requested by @sbailey in an email.  I'm copying it here to make it a bit easier to track the progress.

> Could you create a test script that would do something like:
>
> Starting from a plain edison/cori login (no desi_environment), set $DESI_ROOT to somewhere like $SCRATCH/desi and then proceed to desiInstall a full suite of DESI code needed for the integration test, e.g. desiutil, desitarget, desispec, desisim, specsim, speclite, specter, redmonster.  Then run "python desispec/py/desispec/test/integration_test.py" to verify that all pieces are installed and working as needed.
> 
> That wouldn't actually capture the strange MPI/pmi errors that I have seen when running in parallel, but it is at least a minimal test that all packages are to 0th order desiInstall compliant.
> 
> And then do the same for an anaconda-based distribution, using conda install and git clone + "python setup.py install" (or develop) instead?  This could be used as documentation for how a new user could setup their laptop.
> 
> I expect that these tests will run into problems, which will highlight things that either need to change in the packages themselves or in desiInstall.  We could (should?) turn these into automated tests so that if someone breaks the standard methods of installation for a package, the tests would catch it.  This might be a good use of the Jenkins framework at NERSC.  Maybe.
> 
> Outside of hpcports, NERSC does have anaconda as a python/2.7-anaconda module.  It should be possible to create conda environments with that, upgrade numpy, etc just as one would do on a laptop.

- [x] Easily enable/disable `desi_environment.c?sh` when logging in to NERSC, or identify an account (desi or desiproc) that will be specifically set aside to do this.
- [x] Determine the best *order* in which to install DESI products.
- [x] Script working with desiInstall.
- [x] Test NERSC's existing anaconda environment.
- [x] Script working with anacoda + `pip install git+https://github.com/desihub/...`
- [x] [Ask NERSC Consulting about Jenkins environment.](https://nersc.service-now.com/nav_to.do?uri=incident.do%3Fsys_id=26cbf1266f6a520086e96bbeae3ee410%26sysparm_stack=incident_list.do%3Fsysparm_query=active=true)
- [x] Integrate with Jenkins framework, if available.

**Comment by weaverba137** [comment]
Looks like the 'desi' pseudo-user account might be good for this.

**Comment by weaverba137** [comment]
'desi' account it will be.

**Comment by weaverba137** [comment]
Note: when you push to GitHub, the changed file does not propagate instantaneously to all of GitHub's load-balancing servers.  That is, you can push a commit of a file, but it may take ~1-2 minutes for a subsequent download of that file to reflect that commit.

**Comment by weaverba137** [comment]
In order to run the integration test, clearly packages should be installed in the same order they are installed in the existing integration test.

**Comment by weaverba137** [comment]
I can now run a successful integration test using desiInstall for everything (except redmonster).  Versions used:

desiutil 1.4.0
desitree 0.2.0
desimodules 0.9.3
speclite v0.4
specter 0.4.1
specsim v0.4
desimodel 0.4.4
desitarget 0.3.3
desispec 0.5.0
desisim master (see desihub/desisim#119)

**Comment by weaverba137** [comment]
@sbailey how do you want to proceed with unifying the integration test infrastructure I've developed so far with cron_dailytest.sh?  The primary new features are:

- Start from an absolutely "bare" environment.
- Load arbitrary versions of packages, not just master.
- I figured out how to install bitbucket tags, which turns out to be moderately annoying.

**Comment by weaverba137** [comment]
I have a version of the integration test that "works", modulo properly installing specter/desispec/desitarget(?) data files via pip.

**Comment by sbailey** [comment]
Do they work via git checkout + python setup.py install ?

**Comment by weaverba137** [comment]
No pip install.  But ultimately that's the exact same thing.

**Comment by weaverba137** [comment]
Based on conversations this week, the timescale for getting Jenkins working at NERSC may be longer than the amount of time we want to keep this ticket open.

I propose that we (I):

1. Complete the integration of the integration tests.
2. Complete the install documentation.
3. Close this issue.
4. Open a separate issue for tracking Jenkins development.

**Comment by weaverba137** [comment]
The install documentation (desihub/desispec#170) is very nearly complete.  I have opened a new ticket to specifically track merging the various integration test scripts (desihub/desispec#224).
